LFS-1053: Form actions menu & LFS-1080: As a developer, I can render a Delete button with a text label

veronikaslc opened the pull request:
Jira links:
* [LFS-1053: Form actions menu](https://phenotips.atlassian.net/browse/LFS-1053)
* [LFS-1080: As a developer, I can render a Delete button with a text label](https://phenotips.atlassian.net/browse/LFS-1080)

Also fixes:
* [LFS-1090: Error "404 Not Found" when deleting form while in edit mode](https://phenotips.atlassian.net/browse/LFS-1090)

Commit:
LFS-1053: Form actions menu

sashaandjic commented:
Looks great. One small suggestion: for `Change subject` action use an icon other than pencil.
![image](https://user-images.githubusercontent.com/23243244/113979379-2a839880-9813-11eb-9304-90219bda39bd.png)
![image](https://user-images.githubusercontent.com/23243244/113979188-f019fb80-9812-11eb-80ef-a015d503e6d1.png)
![image](https://user-images.githubusercontent.com/23243244/113979875-d1683480-9813-11eb-8569-a21b08d53c4b.png)
![image](https://user-images.githubusercontent.com/23243244/113980491-8b5fa080-9814-11eb-906f-5f0251f8c3b4.png)
![image](https://user-images.githubusercontent.com/23243244/113979225-f90acd00-9812-11eb-8e74-8fe821f40d5c.png)
![image](https://user-images.githubusercontent.com/23243244/113980251-4176ba80-9814-11eb-89fb-218acf19b76e.png)
![image](https://user-images.githubusercontent.com/23243244/113979499-51da6580-9813-11eb-8ae0-f6803682cdbe.png)
![image](https://user-images.githubusercontent.com/23243244/113980573-a7634200-9814-11eb-81a1-6fd47afab2ce.png)

marta- reviewed (changes requested):
The jira task specifies that the More actions entry "opens a dropdown menu with the options "Change subject" and "Delete" (no icons)". There are 2 main reasons behind this choice:
* the dropdown looks unpolished with just icons as items
* as @sashaandjic pointed out, the current icon we use for "Change subject" is inappropriate (identical to the edit icon), and IMO there isn't a Material UI icon that can express this action clearly without the use of a label.

This means introducing the ability to render the `DeleteButton` as just a label. I suggest adding a new prop `variant` with the possible values "icon", "text", "extended" (extended means both icon and text). I created https://phenotips.atlassian.net/browse/LFS-1080 with more details. Its implementation should be committed to this branch since the More actions dropdown menu is currently the only use case.

Commit:
LFS-1053: Form actions menu
Display actions menu as words instead of icons

Commit:
LFS-1080: As a developer, I can render a Delete button with a text label

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -22,12 +22,14 @@ import { withRouter, useHistory } from "react-router-dom";
 import { Tooltip, Typography, withStyles } from "@material-ui/core";
 import { Delete, Close } from "@material-ui/icons";
+import DeleteIcon from '@material-ui/icons/Delete';
 
```
The  Delete icon is already imported on the line just above

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -22,12 +22,14 @@ import { withRouter, useHistory } from "react-router-dom";
 
 /**
  * A component that renders an icon to open a dialog to delete an entry.
+ * variant - defines the action display; possible values: "icon" (default), "text", "extended" (means both icon and text)
```
Incomplete documentation can be more confusing than no documentation at all

marta- replied:
Please specify the variant prop using prop types and defaults at the bottom.

Commit:
LFS-1080: As a developer, I can render a Delete button with a text label
Addressed codereview comments

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -339,8 +339,16 @@ const questionnaireStyle = theme => ({
-    titleButton: {
-        float: "right"
+    actionsMenu: {
+        border: "1px solid #ddd",
```
Use `theme` color:
```suggestion
        border: "1px solid " + theme.palette.divider,
```

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -169,13 +170,25 @@ function DeleteButton(props) {
+DeleteButton.defaultProps = {
+  variant: "icon" // defines the action display; possible values: "icon" (default), "text", "extended" (means both icon and text)
+}
+
```
```suggestion
DeleteButton.propTypes = {
  variant: PropTypes.oneOf(["icon", "text", "extended"]), // "extended" means both icon and text
}

DeleteButton.defaultProps = {
  variant: "icon",
}

```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -339,8 +339,16 @@ const questionnaireStyle = theme => ({
+        marginTop: theme.spacing(1)
+    },
+    actionsMenuItem: {
+        padding: theme.spacing(0,1,0,1)
```
Let's make this look more like a menu"
```suggestion
        padding: theme.spacing(0,1),
        "& .MuiButtonBase-root" : {
          fontWeight: "normal",
          justifyContent: "flex-start",
          textTransform: "none",
          width: "100%",
       }

```

veronikaslc committed:
LFS-1053: Form actions menu

Use theme color

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-1053: Form actions menu

Better use of default props

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

veronikaslc committed:
LFS-1053: Form actions menu

Optimised CSS styling

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

marta- committed:
LFS-1053: Form actions menu
Better codestyle and preserve URL anchor

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/114086345-7e789680-9880-11eb-811a-409f9cfa2d35.png)

We should do the same on the chart. At least replace gray pencil with the red one.
![image](https://user-images.githubusercontent.com/23243244/114086744-02cb1980-9881-11eb-8d82-0107ca33805d.png)

sdumitriu reviewed (commented):
I would like to make it a practice that whenever we add a composite (list of items) UI element, we add it as an extension point + extensions instead of hardcoding the list in code.

marta- commented:
> I would like to make it a practice that whenever we add a composite (list of items) UI element, we add it as an extension point + extensions instead of hardcoding the list in code.

Created [LFS-1084: Refactor the Form actions menu as an extention point](https://phenotips.atlassian.net/browse/LFS-1084) to address this. I agree this should be a practice. But for now this is a huge user experience improvement and IMO should be merged as is.

marta- commented:
Switched to vertical ellipsis as icon for "More actions" in the latest commit 1f60b0c :

![image](https://user-images.githubusercontent.com/651980/114215608-c99fb180-9933-11eb-96ca-b0e726e093b9.png)

marta- committed:
LFS-1053: Form actions menu
Use vertical ellipsis as icon for "More actions"

sashaandjic commented:
tested 
![image](https://user-images.githubusercontent.com/23243244/114217470-34ea8300-9936-11eb-8659-df7b18575a98.png)

marta- committed:
LFS-1080: As a developer, I can render a Delete button with a text label
* Added label property to DeleteButton, allowing to override the default text/tooltip "Delete <entryType>"
* Renamed onClick handler from handleIconClicked to handleClick, more appropriate since the introduction of text buttons

marta- committed:
LFS-1053: Form actions menu
Improved label and dialog texts for the form's delete button

sashaandjic commented:
Not sure f the following things are related to this PR:
1. when creating a new form the console error "Invalid DOM property 'tabindex'. Did you mean 'tabIndex'?"
2. when deleting form from the patient chart 404 err happens in the following situation:
- create form for subject A
- click on `Save and close`
- click on `Delete form` menu option

If I do the following, 404 err does not happen:
- create a form for subject A
- click on `Save and close`
- click on `Subject` main menu option
- click on subject A
- click on the newly created form
- click on `Delete form` menu option

marta- commented:
> Not sure if the following 2 things are related to this PR:
> 
> 1. when creating a new form the console error` "Invalid DOM property 'tabindex'. Did you mean 'tabIndex'?"`

Unrelated, introduced in #579, and I see it's getting fixed in #589: https://github.com/ccmbioinfo/lfs/pull/589/commits/fb925216109209b5a8ac3b9c88b8ff0b0356183c#diff-6f45c28ff73937e5746dfa2984c3925a8fa7d004fd85b7ccb3b3589cd3b5fb2bR339

marta- committed:
LFS-1053: Form actions menu
LFS-1090: Error "404 Not Found" when deleting form while in edit mode
Redirecting to subject chart instead of previous page after the deletion of a form.

sashaandjic reviewed (approved):
tested

marta- committed:
LFS-1053: Form actions menu
Minor refactoring to avoid duplication

sdumitriu reviewed (commented):
Two minor comments:

- are we sure we want the pencil icon to be red?
- on a form, expand the menu, change the subject -> the menu is still expanded while changing the subject and after the change; I think it should close, if easy to implement.

marta- commented:
> Two minor comments:
> 
> * are we sure we want the pencil icon to be red?

I don't have a strong opinion either way. We can make it primary blue.

> * on a form, expand the menu, change the subject -> the menu is still expanded while changing the subject and after the change; I think it should close, if easy to implement.

Agreed.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -348,22 +365,59 @@ function Form (props) {
+                  </Tooltip>
+                  :
+                  <Tooltip title="Edit">
+                    <IconButton color="secondary" onClick={onEdit}>
```
```suggestion
                    <IconButton color="primary" onClick={onEdit}>
```

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -348,22 +365,59 @@ function Form (props) {
+                >
+                  <List>
+                    <ListItem className={classes.actionsMenuItem}>
+                      <Button onClick={() => {setSelectorDialogOpen(true)}}>
```
I haven't tested, but this may "close" the menu when clicking on "Change subject".
```suggestion
                      <Button onClick={() => {setSelectorDialogOpen(true); setActionsMenu(null)}}>
```

marta- committed:
LFS-1053: Form actions menu

Apply suggestions from code review

sashaandjic reviewed (approved):
re-tested

sdumitriu reviewed (approved):
The delete menu entry still leaves the menu open, but that's not a big deal.
